# Global Electronics Retail Analytics

## Business Problem

A global electronics retailer operates across multiple countries, stores, customers, and product categories.

The objective of this project is to:

- Analyze sales performance
- Evaluate store efficiency
- Identify top-performing products
- Track revenue growth trends
- Generate business recommendations using data-driven insights

## Tools Used

- Python
- Pandas
- NumPy
- Jupyter Notebook

## Dataset

The analysis combines:

- Sales
- Customers
- Products
- Stores
- Exchange Rates

Data Pre-processing and Dataset Overview shown below:


In [1]:
#Write your code here
import pandas as pd
import warnings 
warnings.filterwarnings('ignore')

### Loading each dataset into a variable


In [ ]:
sales_data=pd.read_csv("D:\\UPGRAD\\ASSIGNMENT_FILES\\Assignment_1\\Sales.csv")
customer_info=pd.read_csv("D:\\UPGRAD\\ASSIGNMENT_FILES\\Assignment_1\\Customers.csv",encoding='latin1')
product_info=pd.read_csv("D:\\UPGRAD\\ASSIGNMENT_FILES\\Assignment_1\\Products.csv")
exchange_rates=pd.read_csv("D:\\UPGRAD\\ASSIGNMENT_FILES\\Assignment_1\\Exchange_rates.csv")
stores_info=pd.read_csv("D:\\UPGRAD\\ASSIGNMENT_FILES\\Assignment_1\\Stores.csv")

### Shape of dataset (rows, columns), that's useful 

In [ ]:
print(sales_data.shape)
print(customer_info.shape)
print(product_info.shape)
print(exchange_rates.shape)
print(stores_info.shape)

(62884, 9)
(15266, 10)
(2517, 10)
(11215, 3)
(67, 5)


### Next we check for the Data Types for columns


In [ ]:
print(sales_data.dtypes)
print(customer_info.dtypes)
print(product_info.dtypes)
print(exchange_rates.dtypes)
print(stores_info.dtypes)

Order Number      int64
Line Item         int64
Order Date       object
Delivery Date    object
CustomerKey       int64
StoreKey          int64
ProductKey        int64
Quantity          int64
Currency Code    object
dtype: object
CustomerKey     int64
Gender         object
Name           object
City           object
State Code     object
State          object
Zip Code       object
Country        object
Continent      object
Birthday       object
dtype: object
ProductKey         int64
Product Name      object
Brand             object
Color             object
Unit Cost USD     object
Unit Price USD    object
SubcategoryKey     int64
Subcategory       object
CategoryKey        int64
Category          object
dtype: object
Date         object
Currency     object
Exchange    float64
dtype: object
StoreKey           int64
Country           object
State             object
Square Meters    float64
Open Date         object
dtype: object


### Detailed info: non-null counts, dtypes, memory usage

In [ ]:

sales_data.info()
customer_info.info()
exchange_rates.info()
product_info.info()
stores_info.info()

#Now we check for null values in each dataset
print(customer_info.isnull().sum().sum())
print(sales_data.isnull().sum().sum())
print(exchange_rates.isnull().sum().sum())
print(product_info.isnull().sum().sum())
print(stores_info.isnull().sum().sum())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 62884 entries, 0 to 62883
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   Order Number   62884 non-null  int64 
 1   Line Item      62884 non-null  int64 
 2   Order Date     62884 non-null  object
 3   Delivery Date  13165 non-null  object
 4   CustomerKey    62884 non-null  int64 
 5   StoreKey       62884 non-null  int64 
 6   ProductKey     62884 non-null  int64 
 7   Quantity       62884 non-null  int64 
 8   Currency Code  62884 non-null  object
dtypes: int64(6), object(3)
memory usage: 4.3+ MB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15266 entries, 0 to 15265
Data columns (total 10 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   CustomerKey  15266 non-null  int64 
 1   Gender       15266 non-null  object
 2   Name         15266 non-null  object
 3   City         15266 non-null  object
 4   State Cod

### Data Quality Assessment

The following issues were identified:

- Date columns stored as object data types
- Currency fields stored as strings
- Missing values present in customer and store datasets
- Numeric fields requiring datatype conversion

These issues are addressed in the following codes, before Exploratory DATA Analysis.


Converting all date-related columns to a standardised datetime format. 

In [ ]:
#For sales dataset
print(pd.to_datetime(sales_data['Order Date']))
print(pd.to_datetime(sales_data['Delivery Date']))
#For Customers Dataset
print(pd.to_datetime(customer_info['Birthday']))
#For Exchange_Rates dataset
print(pd.to_datetime(exchange_rates['Date']))
#For Stores dataset
print(pd.to_datetime(stores_info['Open Date']))

0       2016-01-01
1       2016-01-01
2       2016-01-01
3       2016-01-01
4       2016-01-01
           ...    
62879   2021-02-20
62880   2021-02-20
62881   2021-02-20
62882   2021-02-20
62883   2021-02-20
Name: Order Date, Length: 62884, dtype: datetime64[ns]
0              NaT
1       2016-01-13
2       2016-01-13
3       2016-01-12
4       2016-01-12
           ...    
62879          NaT
62880   2021-02-24
62881   2021-02-23
62882   2021-02-23
62883   2021-02-23
Name: Delivery Date, Length: 62884, dtype: datetime64[ns]
0       1939-07-03
1       1979-09-27
2       1947-05-26
3       1957-09-17
4       1965-11-19
           ...    
15261   1936-03-25
15262   1992-02-16
15263   1937-11-09
15264   1937-10-12
15265   1965-08-18
Name: Birthday, Length: 15266, dtype: datetime64[ns]
0       2015-01-01
1       2015-01-01
2       2015-01-01
3       2015-01-01
4       2015-01-01
           ...    
11210   2021-02-20
11211   2021-02-20
11212   2021-02-20
11213   2021-02-20
11214   2021-02-2

### Transform object-type columns containing numeric information into numerical data types for analysis 

In [ ]:
print(customer_info['Zip Code'].apply(pd.to_numeric,errors='coerce'))

product_info['Unit Cost USD'] = product_info['Unit Cost USD'].replace('[$,]', '', regex=True)
product_info['Unit Price USD'] = product_info['Unit Price USD'].replace('[$,]', '', regex=True)

print(product_info[['Unit Cost USD','Unit Price USD']].apply(pd.to_numeric,errors='coerce'))


0         5523.0
1         6522.0
2         3380.0
3         5223.0
4         3698.0
          ...   
15261    77017.0
15262    22101.0
15263    28405.0
15264    92501.0
15265    48302.0
Name: Zip Code, Length: 15266, dtype: float64
      Unit Cost USD  Unit Price USD
0              6.62           12.99
1              6.62           12.99
2              7.40           14.52
3             11.00           21.57
4             11.00           21.57
...             ...             ...
2512          43.07          129.99
2513          43.07          129.99
2514           1.71            3.35
2515           1.71            3.35
2516           1.71            3.35

[2517 rows x 2 columns]


### We also check for Duplicates in our dataset, so that we can declare our dataset to be unique


In [ ]:
print(customer_info.duplicated().sum())
print(sales_data.isnull().sum())
print(exchange_rates.isnull().sum())
print(product_info.isnull().sum())
print(stores_info.isnull().sum())


0
Order Number         0
Line Item            0
Order Date           0
Delivery Date    49719
CustomerKey          0
StoreKey             0
ProductKey           0
Quantity             0
Currency Code        0
dtype: int64
Date        0
Currency    0
Exchange    0
dtype: int64
ProductKey        0
Product Name      0
Brand             0
Color             0
Unit Cost USD     0
Unit Price USD    0
SubcategoryKey    0
Subcategory       0
CategoryKey       0
Category          0
dtype: int64
StoreKey          0
Stores_country    0
Stores_state      0
Square Meters     1
Open Date         0
dtype: int64


### Created a consolidated, analysis-ready dataset (sales_full), containing transactional, customer, product, store, and currency information in a single table.


In [ ]:
sales_data['Order Date']=pd.to_datetime(sales_data['Order Date'])
exchange_rates['Date']=pd.to_datetime(exchange_rates['Date'])
sales_fx =sales_data.merge(
    exchange_rates,
    left_on=['Order Date', 'Currency Code'],
    right_on=['Date', 'Currency'],
    how='left'
)

# Merge master data
sales_full = (
    sales_fx
    .merge(customer_info, on='CustomerKey', how='left')
    .merge(product_info, on='ProductKey', how='left')
    .merge(stores_info, on='StoreKey', how='left')
)

### Revenue Standardization Since transactions occurred across multiple currencies, exchange-rate data was integrated to convert revenue into a common USD format. This enables accurate cross-country comparisons and performance benchmarking.

In [12]:
# Create USD-normalised revenue
product_info['Unit Cost USD'] = product_info['Unit Cost USD'].replace('[$,]', '', regex=True)
product_info['Unit Price USD'] = product_info['Unit Price USD'].replace('[$,]', '', regex=True)
sales_full[['Exchange','Quantity','Unit Price USD','Unit Cost USD']] = sales_full[['Exchange','Quantity','Unit Price USD','Unit Cost USD']].apply(
    pd.to_numeric,
    errors='coerce')
sales_full['Revenue_USD'] = (
    sales_full['Quantity'] *
    sales_full['Unit Price USD'] *
    sales_full['Exchange']
)

### Exploratory Analysis Section below:


## Calculation of Total Revenue of the retail stores

In [ ]:
print(sales_full.columns)
sales_full['Revenue_USD'].sum()

Index(['Order Number', 'Line Item', 'Order Date', 'Delivery Date',
       'CustomerKey', 'StoreKey', 'ProductKey', 'Quantity', 'Currency Code',
       'Date', 'Currency', 'Exchange', 'Gender', 'Name', 'City', 'State Code',
       'Customer_state', 'Zip Code', 'Customer_country', 'Continent',
       'Birthday', 'Product Name', 'Brand', 'Color', 'Unit Cost USD',
       'Unit Price USD', 'SubcategoryKey', 'Subcategory', 'CategoryKey',
       'Category', 'Stores_country', 'Stores_state', 'Square Meters',
       'Open Date', 'Revenue_USD'],
      dtype='object')


55349466.101787


## Finding total revenue (in USD) by Product Categories

In [ ]:
print(sales_full.groupby('Category')['Revenue_USD'].sum().round(2))

Category
Audio                             3147476.96
Cameras and camcorders            6502487.56
Cell phones                       6118492.00
Computers                        19143319.65
Games and Toys                     718602.82
Home Appliances                  10747777.59
Music, Movies and Audio Books     3113971.62
TV and Video                      5857337.90
Name: Revenue_USD, dtype: float64


## Finding Top 5 Best Performing Stores by Revenue

In [ ]:
print(sales_full.groupby(['Stores_country','Stores_state'])['Revenue_USD'].sum().sort_values(ascending=False)
    .head(5).round(2))


Stores_country  Stores_state             
Online          Online                       11431369.95
Canada          Northwest Territories         1756645.28
                Newfoundland and Labrador     1617620.45
United States   Nevada                        1417885.41
                Kansas                        1394738.06
Name: Revenue_USD, dtype: float64



### Finding Top 5 Products sold as per number of units/Quantity sold

In [ ]:
print(sales_full.groupby('Product Name')['Quantity'].sum().sort_values(ascending=False)
    .head(5))


Product Name
WWI Desktop PC2.33 X2330 Black                550
WWI Desktop PC1.80 E1800 White                538
Adventure Works Desktop PC1.60 ED160 Black    521
Adventure Works Desktop PC2.30 MD230 White    521
Adventure Works Desktop PC1.80 ED180 Black    520
Name: Quantity, dtype: int64


### KPI 1: Store Revenue Efficiency

Metric:

Revenue per Square Meter

Formula:

Revenue_USD / Square_Meters

Purpose:

Evaluate how efficiently each store utilizes physical retail space.


In [ ]:
sales_full['revenue_per_Sqft']=(sales_full['Revenue_USD']/sales_full['Square Meters'])

print(sales_full.groupby(['Stores_country', 'Stores_state'])['revenue_per_Sqft'].mean().sort_values(ascending=False))

Stores_country  Stores_state                
France          Mayotte                         3.322692
                Corse                           2.974029
Germany         Saarland                        2.416231
France          Limousin                        2.129687
                Franche-Comté                   2.124733
                Basse-Normandie                 2.037700
Australia       Australian Capital Territory    1.968956
France          La Réunion                      1.923422
                Martinique                      1.912131
Australia       Northern Territory              1.741314
Germany         Freie Hansestadt Bremen         1.630827
United States   Wyoming                         1.122416
Canada          Nunavut                         0.897043
Netherlands     Flevoland                       0.821396
Italy           Enna                            0.796930
United States   Alaska                          0.784056
                South Dakota               

### Observation: 
Looking at the data,Stores in Mayotte, France generate approximately 3.32 USD revenue per square meter on average.
Hence this means that more efficient store space usage will be required to make profit for other low performing regions.
From above analysis it can be derived that, smaller stores generate high revenue and excellent sales performance per square meter.

### KPI 2: Recurring growth Month over Month

Metric:

Recurring Revenue Growth monthly

Formula:

((Current Month Revenue−Previous Month Revenue)/Previous Month Revenue)​×100

Purpose:

Evaluate how efficiently store performance for month after month



In [ ]:
sales_full['Order Date'] = pd.to_datetime(sales_full['Order Date'])
sales_full['Month'] = sales_full['Order Date'].dt.to_period('M')
monthly_revenue = sales_full.groupby('Month')['Revenue_USD'].sum()
mom_growth = (monthly_revenue.pct_change() * 100)
print(mom_growth)
print(monthly_revenue.rolling(3).mean())

Month
2016-01           NaN
2016-02     37.096856
2016-03    -62.399617
2016-04    -66.456627
2016-05    407.493544
              ...    
2020-10    -35.285345
2020-11      4.403873
2020-12    144.273383
2021-01    -19.493135
2021-02      1.613933
Freq: M, Name: Revenue_USD, Length: 62, dtype: float64
Month
2016-01              NaN
2016-02              NaN
2016-03    621533.398513
2016-04    443438.833479
2016-05    337185.849141
               ...      
2020-10    347312.271887
2020-11    295427.602068
2020-12    378152.615960
2021-01    464837.473938
2021-02    550625.016200
Freq: M, Name: Revenue_USD, Length: 62, dtype: float64


### Observation:
We can identify from the above analysis betweeen Jan 2016 to Feb 2021, since the store opened, the performance has not been a stable one. As from jan to feb the revenue increased by 37.09 percent approx, but then revenue dropped by 62.39 percent for the month march. Now the reason could be change in demand of the product and since its a new store opening, it will take some time to gain traction, as its very normal for any store business, as you can see for the month of MAY revenue increased by 407.49 percent, but yes sharp drops are there. So my recommendation to the company would be to invest in stronger marketing campaigns to increase their outreach promote their products and focus on strategies to stabelize revenue generation.


## KPI 3: Store Age Performance

Stores were segmented into:

- 0–5 Years
- 5–10 Years
- 10–15 Years
- 15–20 Years
- 20+ Years

The objective was to evaluate whether mature stores outperform newer locations.

In [19]:
sales_full['Open Date'] = pd.to_datetime(sales_full['Open Date'])

sales_full['Store_Age'] = (
    pd.Timestamp.today().year -
    sales_full['Open Date'].dt.year
)
sales_full['Store_Age_Bucket'] = pd.cut(
    sales_full['Store_Age'],
    bins=[0, 5, 10, 15, 20, 100],
    labels=[
        '0–5 years',
        '5–10 years',
        '10–15 years',
        '15–20 years',
        '20+ years'
    ],
    right=False
)


print(
    sales_full[['Open Date', 'Store_Age', 'Store_Age_Bucket']].head()
)

   Open Date  Store_Age Store_Age_Bucket
0 2015-04-04         11      10–15 years
1 2010-01-01         16      15–20 years
2 2010-01-01         16      15–20 years
3 2010-01-01         16      15–20 years
4 2010-01-01         16      15–20 years


In [ ]:
print(
    sales_full.groupby('Store_Age_Bucket')['Revenue_USD']
    .sum()
    .sort_values(ascending=False).round(2)
)

Store_Age_Bucket
15–20 years    28885656.74
10–15 years    21852471.09
20+ years       2402905.46
5–10 years      2208432.82
0–5 years             0.00
Name: Revenue_USD, dtype: float64


In [ ]:
print(
    sales_full.groupby('Store_Age_Bucket')['Revenue_USD']
    .mean()
    .round(2)
    .sort_values(ascending=False))

Store_Age_Bucket
20+ years      934.26
10–15 years    890.56
5–10 years     881.61
15–20 years    868.25
0–5 years         NaN
Name: Revenue_USD, dtype: float64


### Observation:
Older stores have higher average revenue, whereas Mature stores perform better, Newer stores have high average revenue	hence the stores Expansion strategy is effective. Mid-age stores are underperforming, hence Operational optimization may be needed.